### PR TITLE
add hotpotqa benchmark

### PR DESCRIPTION
**Pull request by LittleSunshineQi**

### CODE REVIEW

File: benchmarks/HOTPOTQA/hotpotqa_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        """Initialization."""
+        super().__init__()
+
+    def _evaluate(self, ) -> Tuple[Dict[Any, Any], Dataset]:
+        """Evaluate the dataset and return the dataset with scores.
+
+        For the HotPotQA dataset, the `short_answers` and `long_answers` which we need to extract.
+
+
+        We use the `answer` as the `gt_answers` to evaluate the string Exact Match correctness and the `gt_supporting_facts` to evaluate the F1.
```
Comment (bugtig6351) [inline benchmarks/HOTPOTQA/hotpotqa_benchmark.py:31]: Too many blank lines.

File: benchmarks/HOTPOTQA/hotpotqa_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        """
+        self.metrics = [AnswerEMCorrectness(ignore_case=False),
+                        AnswerF1Correctness()
+                        ]
+        ground_truths = {
+            "answer_f1": ("supporting_answer", "gt_supporting_facts"),
```
Comment (bugtig6351) [inline benchmarks/HOTPOTQA/hotpotqa_benchmark.py:37]: The usage of the F1 metric is incorrect. Please take another look at the code of `AnswerF1Correctness()`. 

When calculating the F1, the two input strings are split into word-level tokens for computation. In this benchmark, “gt_supporting_facts” is a list of supporting sentences, while “supporting_answer” is a string concatenated from the supporting sentences found by the model. Directly inputting “supporting_answer” and “gt_supporting_facts” into `AnswerF1Correctness()` will first calculate the F1 between the token set of the former and the token set of each sentence in the latter, and then take the maximum value from all the F1 scores.
The correct way is calculating the F1 score at sentence level, comparing sentences and their `sent_id` to  determine whether the model has found the appropriate sentences.

**Review by bugtig6351 — CHANGES_REQUESTED** [review]

**Review by Wenshansilvia — CHANGES_REQUESTED** [review]
Please add readme.md file according to https://github.com/gomate-community/rageval/tree/main/benchmarks/ASQA

File: benchmarks/HOTPOTQA/hotpotqa_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    parser.add_argument("--local_file", type=str, default=None)
+
+    args = parser.parse_args()
+    date = time.strftime("%Y%m%d", time.localtime())
+
+    benchmark = HOTPOTQABenchmark()
+    if args.local_file:
+        data_file = os.path.join(args.output_dir, 'output', args.local_file)
+        print(data_file)
+        print(data_file)
```
Comment (QianHaosheng) [inline benchmarks/HOTPOTQA/hotpotqa_benchmark.py:72]: This line may be redundant code.

File: benchmarks/HOTPOTQA/run.sh [hunk 4]
```diff
@@ -0,0 +1,14 @@
+rageval_dir=$(dirname $(dirname $(dirname $(realpath $0))))
+cd $rageval_dir
+
+echo "Running HotPotQA Benchmark"
+
+python3 benchmarks/HOTPOTQA/hotpot_qa_benchmark.py --output_dir "benchmarks/HOTPOTQA" --remote_split "gpt_3.5_turbo"
+
+# 检查返回状态码
```
Comment (QianHaosheng) [inline benchmarks/HOTPOTQA/run.sh:8]: Annotations should be kept in English.

**Review by QianHaosheng — CHANGES_REQUESTED** [review]

File: benchmarks/HOTPOTQA/README.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+For each "question" in the dataset, there is a brief and precise answer in the "answer" and sentence-level supporting facts required for reasoning in the " supporting_facts", the "level" means the difficulty level of the question, the "context" has  additional context relevant to the question ,"id" is an identifier for the question-answer pair and the field "type" indicates a categorization of the question-answer pair.
+
+
+## 3. Metrics
+
+We adopt the default metrics used in the [HotPotQA paper](https://arxiv.org/abs/1809.09600) as follows:
+
+1. [String Exact Match](../../rageval/metrics/_answer_exact_match.py): for each `answer`, check whether it is present in the predictions.
+2. [F1](../../rageval/metrics/_answer_f1.py): utilize the model's predictions to extract supporting facts, which are identified by the model. Subsequently, we compute the F1 score by comparing these extracted supporting facts with the 'supporting_facts' provided in the dataset。
```
Comment (bugtig6351) [inline benchmarks/HOTPOTQA/README.md:40]: "。"

File: benchmarks/HOTPOTQA/README.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+- `--max_num_examples`: The maximum number of examples used in generate answers.
+- `--max_length`: The maximum number of tokens that can be generated for answering questions.
+- `--output_path`: Directory that the generated answers will be saved.
+- `--model`: The OpenAI GPT model to use, e.g., gpt-3.5-turbo. And result file will be named as `f"{model}.jsonl"`.
+- `--api_key`: Your OpenAI API key.
+
+### 4.2 Evaluation
+
+1. Prepare RAG responses. By default, `hotpotqa_benchmark.py` will download the results of the `gpt_3.5_turbo` model from [our huggingface dataset](https://huggingface.co/datasets/golaxy/rag-bench) and evaluate them. If you wish to evaluate your own results, you can simply attach your predictions as `short_answer` and `supporting_answer`  to each example in the original HotPotQA dataset, similar to what we did in [the file](https://huggingface.co/datasets/golaxy/rag-bench/viewer/hotpot_qa/gpt_3.5-_turbo).
```
Comment (bugtig6351) [inline benchmarks/HOTPOTQA/README.md:69]: please update the result file on `rag-bench`, adding the scores of each metric.

File: benchmarks/HOTPOTQA/hotpotqa_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+            recode = sum(len_support[:idx]) + 1 + sent_id
+            recode_answers.append(str(recode))
+        recode_answers=[' '.join(recode_answers)]
+        data["gt_sent_ids"] = recode_answers
+        return data
+
+    def _evaluate(self) -> Tuple[Dict[Any, Any], Dataset]:
+        """Evaluate the dataset and return the dataset with scores.
+
+        For the HotPotQA dataset, the `short_answer` and `supporting_answer` which we need to extract
```
Comment (bugtig6351) [inline benchmarks/HOTPOTQA/hotpotqa_benchmark.py:40]: For the HotPotQA dataset, we need to extract the `short_answer` and `supporting_answer`.

**Review by bugtig6351 — APPROVED** [review]

**Review by QianHaosheng — APPROVED** [review]
Good job ~

**Review by Wenshansilvia — APPROVED** [review]

**Pull request merged**